[Ng Jia Yuan] iP

ngjiayuan opened the pull request:
### DukeUltraProPlus
> "Why live in the present when you can live in the future?" - Jia Yuan

DukeUltraProPlus is not just another bot. It empowers you. 
### It's *SIC*:
* **S**wift
* **I**ntuitive
* **C**onvenient

To get started,
1. Download it [here](https://www.github.com/ngjiayuan/ip).
2. Run it.
3. Add your tasks.
4. Feel empowered. 💪 

Experience the **future**.

Features:
- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders

If you think you are smart, here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

Commit:
Add Gradle support

ngjiayuan committed:
Level-1

ngjiayuan committed:
Level-2

ngjiayuan committed:
Level-3

ngjiayuan committed:
Level-4

ngjiayuan committed:
A-TextUiTesting

ngjiayuan committed:
Level-5

ngjiayuan committed:
Level-6

ngjiayuan committed:
Complete Level-7

ngjiayuan committed:
Complete Level-8

ngjiayuan committed:
Complete A-MoreOOP

ngjiayuan committed:
Organize classes into packages

ngjiayuan committed:
Add JUnit Tests

ngjiayuan committed:
Add JavaDoc comments

ngjiayuan committed:
Update to follow coding standard

ngjiayuan committed:
Add find functionality

ngjiayuan committed:
Fix merge conflict

ngjiayuan committed:
Merge with branch-Level-9

ngjiayuan committed:
Merge remote-tracking branch 'origin/add-gradle-support'

gycgabriel commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,138 @@
+     */
+    private Task stringToTask(String taskString) {
+        Task result = new Task();
+        String taskType = taskString.substring(0, 3);
+        String taskStatus = taskString.substring(3, 6);
+        String taskDetail = taskString.substring(7);
```
Good structure and naming 👍

gycgabriel commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * Runs Duke until the user inputs a bye command.
+     */
+    public void run() {
+        inSession = true;
+        Parser parser = new Parser(taskList, storage);
+        ui.greet();
+        while (inSession) {
+            try {
+                inSession = parser.parseCommand(ui.scan());
+            } catch (DukeException e) {
+                ui.showError(e);
+            }
+        }
+        ui.exit();
+    }
```
Logical structure and KISS excellently 👍
No deep nesting and very readable 👍

gycgabriel commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,172 @@
+            break;
+        case ("done"): {
+            String[] info = input.split(" ", 2);
+            if (info.length == 1) {
+                throw new DukeException("☹ OOPS!!! " +
+                        "Please enter the task number you would like to mark as done.");
```
'Make the Happy Path prominent' done very well here 👍

gycgabriel commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,138 @@
+            if (dateTimeOfDeadline.length == 1) {
+                result = new Deadline(deadlineDescription[0]
+                        .substring(0, deadlineDescription[0].length() - 2),
+                        LocalDate.parse(dateTimeOfDeadline[0], DateTimeFormatter.ofPattern("dd MMM yyyy")));
```
I find it a bit messy here, can consider adding a line break here and in other places: (purely optional)
```suggestion
                        LocalDate.parse(dateTimeOfDeadline[0], 
                        DateTimeFormatter.ofPattern("dd MMM yyyy")));
```

gycgabriel reviewed (commented):
LGTM, could not find any issues with code quality. I found your code to be very clean and minimalistic, yet managing to accomplish the features required, which is definitely a learning point for me! 👍

ngjiayuan committed:
Add Gradle support

ngjiayuan committed:
Add CheckStyle and correct styling

ngjiayuan committed:
Add GUI to Duke

jyrw commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,204 @@
+    private TextField userInput;
+    private Button sendButton;
+    private Scene scene;
+//    private Image user = new Image(this.getClass()
+//            .getResourceAsStream("/images/DaUser.png"));
+//    private Image duke = new Image(this.getClass()
+//            .getResourceAsStream("/images/DaDuke.png"));
```
Probably should clean up commented out code.

jyrw commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,204 @@
+//        });
+//    }
+
+    /**
+     * Iteration 1:
+     * Creates a label with the specified text and adds it to the dialog container.
+     * @param text String containing text to add
+     * @return a label with the specified text that has word wrap enabled.
+     */
+//    private Label getDialogLabel(String text) {
+//        // You will need to import `javafx.scene.control.Label`.
+//        Label textToAdd = new Label(text);
+//        textToAdd.setWrapText(true);
+//
+//        return textToAdd;
+//    }
+
+    /**
+     * Iteration 2:
+     * Creates two dialog boxes, one echoing user input and the other containing Duke's reply and then appends them to
+     * the dialog container. Clears the user input after processing.
+     */
+//    private void handleUserInput() {
+//        Label userText = new Label(userInput.getText());
+//        Label dukeText = new Label(getResponse(userInput.getText()));
+//        dialogContainer.getChildren().addAll(
+//                DialogBox.getUserDialog(userText, new ImageView(user)),
+//                DialogBox.getDukeDialog(dukeText, new ImageView(duke))
+//        );
+//        userInput.clear();
+//    }
```
Not sure if it's necessary to show iterations here if it's in commit history.

jyrw commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,204 @@
+    /**
+     * Initialises storage, taskList and ui.
+     *
+     * @param filePath the relative filePath for storage of dataFile.
```
I think Javadoc @param/@return tags' descriptions should be capitalized based on [this](https://se-education.org/guides/conventions/java/intermediate.html#comments), but maybe it's just preference.

jyrw reviewed (commented):
Looks good. Really extremely difficult to find any violations to comment on. Some of the recently added classes for the GUI are missing some Javadocs.

ngjiayuan committed:
Use assertions

Document important assumptions in the code.

Fix if-else statements to reflect assumptions and use assertions in
unreachable else and default blocks.

ngjiayuan committed:
Merge pull request #1 from ngjiayuan/branch-A-Assertions

Use assertions

ngjiayuan committed:
Improve code quality

To examine and improve code quality for readability.

Fix long methods and fix complicated expressions.

ngjiayuan committed:
Merge pull request #2 from ngjiayuan/branch-A-CodeQuality

Improve code quality

ngjiayuan committed:
Use Streams

ngjiayuan committed:
Merge pull request #3 from ngjiayuan/branch-A-Streams

Use Streams

ngjiayuan committed:
Set up CI

ngjiayuan committed:
Add reminder functionality

Provide a way for user to get reminders about tasks.

Add a new command 'remind' to get the tasks for a certain period.
For example, 'remind today' will return the tasks for today.

Adding a period will allow users to see all the tasks for the period
and be more focused.

The supported commands are 'remind today', 'remind tomorrow',
'remind week' and 'remind month'.

ngjiayuan committed:
Add user guide and screenshot

ngjiayuan committed:
Edit build.gradle

ngjiayuan committed:
Fix linkage error
